Feat nrf time rtc support

hoihu opened the pull request:
This is a cleanup version of the inputs collected in https://github.com/micropython/micropython/pull/6171
(I didn't want to force push the update to that PR because it contains implementation details
via ticker module that might be useful for others)

Basically it adds support for time.ticks_ms / time.ticks_us via RTC1, configured in 30usec tick mode. It also 
adds the time.ticks_add and time.ticks_diff helper functions.

For this to work properly the RTC1's overflow irq is handled and a msec / usec ticks counter appropiately 
generated (see also commit comment for this relatively complex matter). Thanks for @dpgeorge for 
basically providing the solution there!

The PR also contains support for MICROPY_EVENT_POLL_HOOK. The wakeup from WFI is done via RTC1's 
CCR0, configured in app. 1msec intervals. Basically that makes the nrf port ready for asyncio.

This feature can be enabled using MICROPY_PY_TIME_TICKS. If disabled, the system uses the legacy 
ticks and sleep methods.

dpgeorge commented on ports/nrf/modules/machine/rtcounter.c:
```diff
@@ -30,6 +30,7 @@
 #include "rtcounter.h"
 #include "nrfx_rtc.h"
 #include "nrf_clock.h"
+#include "py/mperrno.h"
```
I guess this addition is not needed

hoihu replied:
ok

dpgeorge commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,113 @@
+    }
+    // irq handler for wakeup from WFI (~1msec)
+    if (event == NRFX_RTC_INT_COMPARE0) {
+        nrfx_rtc_cc_set(&rtc1, 0, nrfx_rtc_counter_get(&rtc1) + 33, true);
```
the "33" should probably be a #define constant

hoihu replied:
yes

dpgeorge commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,113 @@
+    nrfx_rtc_init(&rtc1, &rtc_config_time_ticks, rtc_irq_time);
+    nrfx_rtc_overflow_enable(&rtc1, true);
+    // generate a ~1msec irq to leave _WFI
+    nrfx_rtc_cc_set(&rtc1, 0, 30, true);
```
the "30" should be the same as the "33" above, right?  in which case use the same #define constant

hoihu replied:
that would be the first CC irq to schedule, so can actually be different? 

But I haven't looked at it closely. It seems that nrfx_rtc_init does not clear the counter state (as I have assumed). At least I didn't find any reference to it.

So I believe we should reset the counter using `nrfx_rtc_counter_clear`, as pointed below from @glennrub  of use a similar logic like `nrfx_rtc_counter_get(&rtc1) + xxx` ? My preference would be the second one.

dpgeorge replied:
> My preference would be the second one.

Sounds good.

dpgeorge commented on ports/nrf/mphalport.c:
```diff
@@ -70,7 +177,11 @@ int mp_hal_stdin_rx_chr(void) {
         }
+        #if MICROPY_PY_TIME_TICKS
+        MICROPY_EVENT_POLL_HOOK
+        #else
```
could it just use `MICROPY_EVENT_POLL_HOOK` unconditionally (ie in both cases here)?

hoihu replied:
Yes I was thinking about that too.. (since there was already a WFI statement). But I wanted it to behave like the old, legacy solution if `MICROPY_PY_TIME_TICKS` was disabled. But I can easily enable it always, no problem.

dpgeorge replied:
aah, since `MICROPY_ENABLE_SCHEDULER` is disabled on nrf it won't make a difference, `mp_handle_pending()` won't do anything except check for pending exceptions, but at the REPL (where this function is used) ctrl-C (the only possible pending exception) is disabled

hoihu replied:
ok

dpgeorge commented:
Looks really good!

glennrub commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,113 @@
+    if (!nrf_clock_lf_is_running(NRF_CLOCK)) {
+        nrf_clock_task_trigger(NRF_CLOCK, NRF_CLOCK_TASK_LFCLKSTART);
+    }
+    nrfx_rtc_init(&rtc1, &rtc_config_time_ticks, rtc_irq_time);
```
Should we add a `nrfx_rtc_counter_clear()` here? Right now, it remembers the old counter value across `soft reset`. Or is it more interesting to have the count from power on?

dpgeorge replied:
I think just leave it, to keep it simple.  Ticks should anyway not be used as an absolute measure of time, just for relative differences (that's how it works in most other ports, it's only reset on power-up).

glennrub replied:
I agree. I also noticed the documentation after writing the question which says: "... counter with an arbitrary reference point ...".

hoihu replied:
> Should we add a nrfx_rtc_counter_clear() here?

see comment above

glennrub commented:
From my side, it looks good as well.
I've tested the patch a bit, including WFI, overflow and my obscure "IRQ in IRQ" scenario. I all seems to work very well. I've also test integrated it into my nrf91 sockets, and you have solved my need for timer. Thank you @hoihu !

hoihu commented:
> From my side, it looks good as well.
I've tested the patch a bit, including WFI, overflow and my obscure "IRQ in IRQ" scenario. I all seems to work very well. I've also test integrated it into my nrf91 sockets, and you have solved my need for timer. Thank you @hoihu !

Great, thanks!

hoihu committed:
ports/nrf: add time module ticks support.

Add time.ticks_ms / time.ticks_us support using RTC1 as timebase.
Also add the time.ticks_add resp. time.ticks_diff helper functions.

This feature can be enabled using MICROPY_PY_TIME_TICKS. If disabled
the system uses the legacy sleep methods.

In addition support for MICROPY_EVENT_POLL_HOOK was added
to the time.sleep_ms(x) function, making this function more power
efficient and allows support for select.poll / asyncio. To support
this, the RTC's CCR0 was used to schedule a ~1msec event to wakeup
the CPU.

Some important notes about the RTC timebase:

- Since the granularity of RTC1's ticks are app. 30usec, time.ticks_us
is not perfect, but moreless usable. For tighter measurments, the
ticker's 1MHz counter should be used.

- time.ticks_ms(x) should *not* be called in an irq with
higher prio than the RTC overflow irq (3). If so, it introduces
a race condition and possibly leads to wrong tick calculations.

- See also https://github.com/micropython/micropython/pull/6171
for further informations.

hoihu committed:
ports/nrf: enable nrf tick support.

hoihu commented:
I force pushed the review feedbacks, changes are:
- `MICROPY_EVENT_POLL_HOOK` is enabled unconditionally in `mp_hal_stdin_rx_chr`
- The first RTC's CC irq is now properly scheduled (I used a macro there too). The RTC counter is soft-reset proof (not sure if that really matters, but might be nice on the REPL).
- Deleted unused #include

From my point of view ready to go in.

dpgeorge commented:
Thanks for updating, and for the effort putting this together.  Now merged in 15574cd665a8e66157aa97447e3869513da81d45 and 59ed3bdd9fa8dc6e0677cf8d4478a5a32fee4500 (with minor edits to reorder some bits, and remove `MICROPY_PY_TIME_TICKS` guard from `MICROPY_EVENT_POLL_HOOK` definition).
